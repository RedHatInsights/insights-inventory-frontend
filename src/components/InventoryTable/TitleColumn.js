/* eslint-disable react/prop-types */
import React from 'react';
import { Link } from 'react-router-dom';
import { ConversionPopover } from './ConversionPopover/ConversionPopover';
import { Icon, Popover } from '@patternfly/react-core';
import { BundleIcon } from '@patternfly/react-icons';
import FontAwesomeImageIcon from '../FontAwesomeImageIcon';

/**
 * Helper function to proprly calculate what to do when user clicks on first cell.
 * Either full redirect if used with ctrl button or `onRowClick` from props is used.
 * @param {*} event html event, to find out if meta key was clicked.
 * @param {*} key inventory UUID.
 * @param {*} props additional props from `EntityTable` - loaded, onRowClick and noDetail.
 */
const onRowClick = (event, id, { loaded, onRowClick: rowClick, noDetail }) => {
  if (loaded && !noDetail) {
    const isMetaKey = event.ctrlKey || event.metaKey || event.which === 2;
    if (isMetaKey) {
      return;
    } else if (rowClick) {
      rowClick(event, id, isMetaKey);
    }
  }

  event.preventDefault();
  event.stopPropagation();
};

/**
 * Helper component to generate first cell in plain inventory either with clickable detail or just data from attribut.
 * This is later on used in redux in `renderFunc`.
 * @param {React.node} children React node with information that will be shown to user as column title.
 * @param {string} id inventory UUID, used to navigate to correct URL.
 * @param {*} item row data, holds every information from redux store for currecnt row.
 * @param {*} props additional props passed from `EntityTable` - holds any props passed to inventory table.
 */
const TitleColumn = ({ children, id, item, ...props }) => (
  <div className="ins-composed-col sentry-mask data-hj-suppress">
    {item?.os_release && <div key="os_release">{item?.os_release}</div>}
    <div key="data" className={props?.noDetail ? 'ins-m-nodetail' : ''}>
      {props?.noDetail ? (
        children
      ) : (
        <span>
          {item?.system_profile?.bootc_status ? (
            <Popover
              triggerAction="hover"
              headerContent="Image-based system"
              bodyContent={
                <div>
                  Image mode for Red Hat Enterprise Linux is a container-native
                  approach that uses the same bits but delivers them as a
                  container image. Updates are immutable and the experience is
                  very close to running a containerized application.
                </div>
              }
            >
              <Icon style={{ marginRight: '8px' }}>
                <FontAwesomeImageIcon
                  fill="var(--pf-v5-global--icon--Color--light)"
                  margin="0px"
                />
              </Icon>
            </Popover>
          ) : (
            <Popover
              triggerAction="hover"
              headerContent="Package-based system"
              bodyContent={
                <div>
                  Package mode is a familiar RHEL experience across any
                  footprint where the OS is assembled and updated from rpm
                  packages. This is traditionally how RHEL is deployed and will
                  remain the preferred method for many.
                </div>
              }
            >
              <Icon style={{ marginRight: '8px' }}>
                <BundleIcon color="var(--pf-v5-global--icon--Color--light)" />
              </Icon>
            </Popover>
          )}
          <Link
            to={item?.href || item?.to || id}
            {...{
              ...(props?.onRowClick
                ? {
                    onClick: (event) => {
                      onRowClick(event, id, props);
                    },
                  }
                : {}),
            }}
          >
            {children}
          </Link>
          {item?.system_profile?.operating_system?.name === 'CentOS Linux' && (
            <ConversionPopover />
          )}
        </span>
      )}
    </div>
  </div>
);

export default TitleColumn;
