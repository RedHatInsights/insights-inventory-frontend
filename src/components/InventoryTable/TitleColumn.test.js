import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import TitleColumn from './TitleColumn';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  // eslint-disable-next-line react/prop-types
  Link: ({ children, ...props }) => (
    <a className="fakeLink" {...props}>
      {children}
    </a>
  ),
}));

describe('TitleColumn', () => {
  it('should render correctly with NO data', () => {
    const { asFragment } = render(<TitleColumn />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with data', () => {
    const { asFragment } = render(
      <TitleColumn id="testId">something</TitleColumn>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with os_release', () => {
    const { asFragment } = render(
      <TitleColumn id="testId" item={{ os_release: 'os_release' }}>
        something
      </TitleColumn>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with href', () => {
    const { asFragment } = render(
      <TitleColumn id="testId" item={{ href: '/link/to/item' }}>
        something
      </TitleColumn>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly with to', () => {
    const { asFragment } = render(
      <TitleColumn id="testId" item={{ to: { pathname: '/link/to/item' } }}>
        something
      </TitleColumn>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render correctly no detail with data', () => {
    const { asFragment } = render(
      <TitleColumn
        id="testId"
        item={{ os_release: 'os_release' }}
        noDetail={true}
      >
        something
      </TitleColumn>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('API', () => {
    it('should call onClick', async () => {
      const onClick = jest.fn();
      render(
        <TitleColumn
          id="testId"
          item={{ os_release: 'os_release' }}
          onRowClick={onClick}
          loaded
        >
          something
        </TitleColumn>
      );

      await userEvent.click(screen.getByText(/something/i));
      expect(onClick).toHaveBeenCalled();
    });

    it('should not call onClick if not loaded', async () => {
      const onClick = jest.fn();
      render(
        <TitleColumn
          id="testId"
          item={{ os_release: 'os_release' }}
          onRowClick={onClick}
        >
          something
        </TitleColumn>
      );

      await userEvent.click(screen.getByText(/something/i));
      expect(onClick).not.toHaveBeenCalled();
    });
  });
});
