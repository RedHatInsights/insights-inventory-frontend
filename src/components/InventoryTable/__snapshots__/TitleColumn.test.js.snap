// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TitleColumn should render correctly no detail with data 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div>
      os_release
    </div>
    <div
      class="ins-m-nodetail"
    >
      something
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleColumn should render correctly with NO data 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div
      class=""
    >
      <span>
        <a
          class="fakeLink"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleColumn should render correctly with data 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div
      class=""
    >
      <span>
        <a
          class="fakeLink"
          to="testId"
        >
          something
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleColumn should render correctly with href 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div
      class=""
    >
      <span>
        <a
          class="fakeLink"
          to="/link/to/item"
        >
          something
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleColumn should render correctly with os_release 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div>
      os_release
    </div>
    <div
      class=""
    >
      <span>
        <a
          class="fakeLink"
          to="testId"
        >
          something
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TitleColumn should render correctly with to 1`] = `
<DocumentFragment>
  <div
    class="ins-composed-col sentry-mask data-hj-suppress"
  >
    <div
      class=""
    >
      <span>
        <a
          class="fakeLink"
          to="[object Object]"
        >
          something
        </a>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
